src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs

```diff
@@ -293,8 +382,79 @@ private TextSpan GetRegionSpanForReference(SourceText sourceText, TextSpan refer
+                try
+                {
+                    lock (Gate)
+                    {
+                        foreach (var customColumnName in customData.Keys)
+                        {
+                            // Get the matching custom column.
+                            var customColumnDefinition = TableControl.ColumnDefinitionManager.GetColumnDefinition(customColumnName) as AbstractFindUsagesCustomColumnDefinition;
+                            if (customColumnDefinition == null)
+                            {
+                                Debug.Fail($"{nameof(SourceReferenceItem.ReferenceInfo)} has a key '{customColumnName}', but there is no exported '{nameof(AbstractFindUsagesCustomColumnDefinition)}' with this name.");
+                                continue;
+                            }
+
+                            // Ensure that we flip the visibility to true for the custom column.
+                            // Note that the actual UI update happens outside the lock when we
+                            // invoke "TableControl.SetColumnStates" below.
+                            ColumnState2 newColumnStateOpt = null;
+                            if (_customColumnTitleToStatesMap.TryGetValue(customColumnDefinition.Name, out var columnState))
+                            {
+                                if (!columnState.IsVisible)
+                                {
+                                    newColumnStateOpt = new ColumnState2(columnState.Name, isVisible: true, columnState.Width,
+                                        columnState.SortPriority, columnState.DescendingSort, columnState.GroupingPriority);
+                                }
+                            }
+                            else
+                            {
+                                newColumnStateOpt = customColumnDefinition.DefaultColumnState;
+                            }
+
+                            if (newColumnStateOpt != null)
+                            {
+                                _customColumnTitleToStatesMap[customColumnDefinition.Name] = newColumnStateOpt;
+
+                                newColumnStates.Add(newColumnStateOpt);
+                            }
+                        }
+                    }
+
+                    // Update the column states if required.
+                    if (newColumnStates.Count > 0)
+                    {
+                        // SetColumnStates API forces a switch to UI thread, so it should be safe to call
+                        // from a background thread here.
+                        // Also note that we will call it only once for each new custom column to add for
+                        // each find references query - the lock above guarantees that newColumnStatesOpt is
+                        // going to be non-null only for the first result that has a non-empty column value.
+                        TableControl.SetColumnStates(newColumnStates);
```

Comment:
note: this is only safe if SetColumnStates doesn't hold onto newColumnStates.  i was thnking there would still be a .ToImmutable here.

Reply:
Sure.

---
In reply to: [222372159](https://github.com/dotnet/roslyn/pull/30155#discussion_r222372159) [](ancestors = 222372159)